psi4/driver/driver_cbs.py

```diff
@@ -703,6 +893,149 @@ def return_energy_components():
 
 VARH = return_energy_components()
 
+def _get_default_xtpl(nbasis, xtpl_type):
+    """ A helper function to determine default extrapolation type.
+    
+    Parameters
+    ----------
+    nbasis : int
+        Number of basis sets
+    xtpl_type : str
+        Extrapolation type ("scf" or "corl").
+    
+    Returns
+    -------
+    function
+        Extrapolation function to be used.
+    """
+    
+    if nbasis == 1:
+        return xtpl_highest_1
+    elif xtpl_type ==  "scf":
+        if nbasis == 2:
+            return scf_xtpl_helgaker_2
+        elif nbasis == 3:
+            return scf_xtpl_helgaker_3
+        else:
+            raise ValidationError("Wrong number of basis sets supplied to scf_xtpl: %d" % nbasis)
+    elif xtpl_type ==  "corl":
+        if nbasis == 2:
+            return corl_xtpl_helgaker_2
+        else:
+            raise ValidationError("Wrong number of basis sets supplied to corl_xtpl: %d" % nbasis)
+            
+
+def _process_cbs_kwargs(kwargs):
+    """ A helper function which either translates supplied kwargs into the
+    ``cbs_metadata`` format, or validates a provided cbs_metadata array.
+    
+    Parameters
+    ----------
+    kwargs : dict
+        kwargs containing the CBS function specification.
+    
+    Returns
+    -------
+    array
+        Array of dictionaries, with each item consisting of an extrapolation
+        stage. All validation takes place here.
+    """    
+    
+    metadata = []
+    if "cbs_metadata" in kwargs:
+        cbs_metadata = kwargs.get("cbs_metadata")
+        for item in cbs_metadata:
+            metadata.append({})
+            if "wfn" in item:
+                if "wfn" not in item:
+                    raise ValidationError("Wavefunction type has to be supplied for every stage.")
+                else:
+                    metadata[-1]["wfn"] = item.get("wfn")
+                if "basis" not in item:
+                    raise ValidationError("Basis set has to be supplied for every stage.")
+                else:
+                    metadata[-1]["basis"] = _expand_bracketed_basis(item.get("basis").lower())
+                metadata[-1]["treatment"] = item.get("treatment", "scf" if len(metadata) == 1 else "corl")
```

Comment:
I'm not sure I understand. The old behaviour of just specifying "corl_wfn" and higher, without specifying "scf_wfn" still works. For "cbs_metadata", the first item in the array is by default a "scf" stage, which means a scf extrapolation method will be used (as opposed to corl extrapolation) - it can (as before) be an extrapolated, or highest_1 stage, but is not restricted to "scf"/"hf"/"c4-" anymore. The extrapolation (scf vs corl) is overridable by providing the "treatment" key, the name of the stage is overridable by the "stage" key.

Reply:
But is it correct that `energy(cbs, cbs_metadata=[{'wfn': 'mp2', 'basis': 'cc-pv[dt]z'}])` is no longer valid b/c no scf stage? What did you have in mind as something ppl would want other than HF (in various flavors and sources)?

Reply:
So there are a few options:
```
[1] energy("mp2/cc-pv[dt]z+d:ccsd(t)/cc-pvdz")
[2] energy(cbs, corl_wfn="mp2", corl_basis="cc-pv[dt]z", 
                delta_wfn="ccsd(t)", delta_basis="cc-pvdz")
[3] energy(cbs, cbs_metadata=[{"wfn": "hf", "basis": "cc-pvtz"},
                              {"wfn": "mp2", "basis": "cc-pv[dt]z"},
                              {"wfn": "ccsd(t)", "basis": "cc-pvdz"}])
E = hf/cc-pvtz + (mp2 - hf)/cc-pv[dt]z + (ccsd(t) - mp2)/cc-pvtz
E = -76.35375057

[4] energy(cbs, cbs_metadata=[{"wfn": "mp2", "basis": "cc-pv[dt]z"},
                              {"wfn": "ccsd(t)", "basis": "cc-pvdz"}])
E = mp2/cc-pv[dt]z + (ccsd(t) - mp2)/cc-pvdz
E = -76.35033786

[5] energy(cbs, cbs_metadata=[{"wfn": "mp2", "basis": "cc-pv[dt]z", "treatment": "corl"},
                              {"wfn": "ccsd(t)", "basis": "cc-pvdz"}])
E = mp2/cc-pv[dt]z + (ccsd(t) - mp2)/cc-pvdz
E = -76.36620899

```
The first three calls are equivalent and are the default behaviour. The last two differ from one another by calling "scf_xtpl_helgaker_2" in [4] and "corl_xtpl_helgaker_2" in [5] directly over the mp2 energies.

Reply:
I'm separately confused by your `[1--5]`. By my common practice (and hence cbs's), `mp2/cc-pv[dt]z == hf/cc-pvtz + (mp2 - hf)/cc-pv[dt]z` with the last term using scf_xtpl_helgaker_2. So `[1--5]` should all be the same. Are you reasoning differently?

If `[4]` is, in fact, the same as `[1--3]`, then that resolves my original concern. That is, is it necessary to explicitly have SCF as the first item in the `cbs_metadata` list? Apparently not, and all's well.

If it clarifies anything, I don't think one should ever apply a correlation energy extrapolation equation to a total energy, unless compensating by subtracting out the reference energy subjected to the same (unnatural b/c eqn designed for correlation) equation is the next step.

Reply:
Yes, currently `[3]` and `[4]` are different, as in  `[1,2,3]` the energy is taken as:
```
  -------------------------------------------------------------------------------------
      Stage      Method / Basis             Wt      Energy [Eh]   Scheme
  -------------------------------------------------------------------------------------
        scf          hf / cc-pvtz            1     -76.05098620   xtpl_highest_1
       corl         mp2 / cc-pv[dt]z         1     -76.35320638   corl_xtpl_helgaker_2
       corl          hf / cc-pv[dt]z        -1     -76.06344462   corl_xtpl_helgaker_2
     delta1     ccsd(t) / cc-pvdz            1     -76.23906022   xtpl_highest_1
     delta1         mp2 / cc-pvdz           -1     -76.22605761   xtpl_highest_1
  -------------------------------------------------------------------------------------
```
In the fourth case it is:
```
  ------------------------------------------------------------------------------------
      Stage     Method / Basis             Wt      Energy [Eh]   Scheme
  ------------------------------------------------------------------------------------
        scf        mp2 / cc-pv[dt]z         1     -76.33733525   scf_xtpl_helgaker_2
       corl    ccsd(t) / cc-pvdz            1     -76.23906022   xtpl_highest_1
       corl        mp2 / cc-pvdz           -1     -76.22605761   xtpl_highest_1
  ------------------------------------------------------------------------------------
```
and in the fifth case:
```
  ------------------------------------------------------------------------------------
      Stage     Method / Basis             Wt      Energy [Eh]   Scheme
  ------------------------------------------------------------------------------------
        scf        mp2 / cc-pv[dt]z         1     -76.35320638   *corl_xtpl_helgaker_2*
       corl    ccsd(t) / cc-pvdz            1     -76.23906022   xtpl_highest_1
       corl        mp2 / cc-pvdz           -1     -76.22605761   xtpl_highest_1
  ------------------------------------------------------------------------------------
```
If you think allowing `[4,5]` is not a good idea, I can validate the `wfn` key of the first item and raise a validation error if it's not one of `HF/SCF/C4-HF/C4-SCF`.

Reply:
I _do_ think allowing `[4,5]` is a Bad Idea. Though I'd relent if everyone disagrees.

Validating `wfn` of the first item if not `HF/C4-HF` (fine by me if the vague `SCF` gets dropped) is a good plan. My notion of a better plan would be if first item not `HF/C4-HF`, just create `{'wfn': "hf", "treatment": highest_1, 'basis': copy_corl_basis_str}` such that the first iteration through `cbs_metadata` creates _two_ entries, the scf and corl stages. Whichever you prefer.